src/v2/user_account.rs

```diff
@@ -0,0 +1,214 @@
+        }
+
+        error: client::Error
+
+    }
+
+    impl<'a> Request<'a> {
+        /// Creates an `Request` with the given user ID.
+        pub fn new(user_id: &'a UserId) -> Self {
+            Self { user_id }
+        }
+    }
+
+    impl Response {
+        /// Creates a new `Response` with all parameters defaulted.
+        pub fn new() -> Self {
+            Default::default()
+        }
+    }
+}
+
+
+pub mod create_modify_user {
+    //! [GET /_synapse/admin/v2/users/:user_id](https://github.com/matrix-org/synapse/blob/master/docs/admin_api/user_admin_api.rst#create-or-modify-account)
+    use super::*;
+
+    ruma_api! {
+        metadata: {
+            description: "create or modify account endpoint",
+            method: PUT,
+            name: "create_modify_account",
+            path: "/_synapse/admin/v2/users/:user_id",
+            rate_limited: false,
+            authentication: AccessToken,
+        }
+
+        request: {
+            /// user ID for the account to renew
+            #[ruma_api(path)]
+            pub user_id: &'a UserId,
+
+            /// This is an optional parameter. Add this parameter to create an account or set this
+            /// password as new one for an existing account.
+            pub password: Option<String>,
+
+            // NOTE: Server explodes if attributes are not omitted but specified as null, like the default
+            // Serde case.
+
+            /// defaults to user_id, or the current value if user already exists
+            /// Some("") is treated as setting it to null.
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub displayname: Option<String>,
+
+            /// defaults to empty, or the current value if user already exists
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub threepids: Option<Vec<super::ThirdPartyIdentifier>>,
+
+            /// The user's avatar URL, if set.
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub avatar_url: Option<String>,
+
+            /// Should the user be a server admin
+            /// defaults to false, or the current value if user already exists
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub admin: Option<bool>,
+
+            /// Should the user be deactivated
+            /// defaults to false, or the current value if user already exists
+            #[serde(skip_serializing_if="Option::is_none")]
+            pub deactivated: Option<bool>,
+
+        }
+
+        #[derive(Default)]
+        response: {
+            /// Information about the user.
+            #[ruma_api(body)]
+            pub info: UserInformation,
+        }
+
+        error: client::Error
+
+        // todo following to does are from synadminctl
+        // TODO: returns 200 if account-exist-and-was-updated,
+        // but 201 CREATED if a new account was created.
+        // However, ruma does throw away this information.
```

Comment:
Interesting. Yeah, there is no way in `ruma_api!` to distinguish these cases right now. I think unless this becomes a common pattern for Matrix, we should just implement the conversion of `http::Request` to `create_modify_user::Request` manually (https://github.com/ruma/ruma/issues/238 would help with that).